Redesign the language fallback system

mooeypoo opened the pull request:
Fallback on languages now has the following rules, in order,
when checking and prioritizing existing values:

1. Requested language
2. Fallback language from fallbacks.json
3. Local language with suffix of same script as requested language
4. Any language with the same script as requested language
5. English
6. Any language with suffix Latn
7. Local language (name tag)
8. First value given

mooeypoo committed:
[wip] Add script fallbacks if no fallback is found

Add the CLDR data 'script' as a fallback in case no language
fallback is found, and before settling for "en" or local language.

This brings back language_scripts dependency, but requires no
updates to that remote repo.

Please hold merging while we discuss; tests were added to show
what the process looks like. This may mitigate the issues of
semi-random language codes in OSM while still sticking to the
"proper" way of doing language fallback based *primarily* on the
language code and only if no other choice then on the script.

nyurik commented:
@mooeypoo I'm not sure I understand - there is no "Cyrl" language code in OSM, only codes like xx_cyrl (or xx-cyrl -- different for each suffix)

mooeypoo commented:
Okay, I'm a little confused. I looked at the list of codes from OSM and I did see Latn, so I thought those were actual codes.

Are you saying they're potential suffixes? Because if so, I can change the code for that.

nyurik commented:
@mooeypoo take a look at the first few pages of https://taginfo.openstreetmap.org/search?q=name%3A -- the suffixes are either *_rm (roman letters), *-Latn (latin lettelrs), _pinyin (Chinese in latin-like), _kana & -Hira (various other forms that I'm not an expert in).

mooeypoo commented:
Okay, so if we are to use the scripts and overrides that were previously given with the language_code repo -- should we add them to the search fallback as suffixes? What's the best way to do this in a way that can mitigate even a little bit the lack of standardization while trying to follow what they do have, with the information and CLDR-driven data that we do have?

Take a look at the new test suite I've added where it specifies what the fallback list per language is expected to be; I think the **order** of the fallback is theoretically correct (requested language, fallback language, fallback script(s), English, local) -- but the fallback scripts are currently, as you point out, not going to work. So... what's the best way to implement this to get it to work as best as possible? Adding as suffixes? With both '-' and '_' options, to maximize the option that either exists?

nyurik commented:
The suffixes are fairly normalized, e.g.  [-Latn](https://taginfo.openstreetmap.org/search?q=-Latn) is almost never [_Latn](https://taginfo.openstreetmap.org/search?q=_Latn) (I myself sometimes fix those), so no need to guess if it should be -* or _*.  The harder issue is if you are rendering Japan for English speakers, should you use Italian or ja-Latn?

mooeypoo commented:
A lot of those decisions will probably have to (a) be tweaked in the future and (b) require probably heavier consideration for fallback rules per language or region.

We can discuss those, but I think that as a first step, having *some* fallback to a script is viable. I don't know what the normal rules are with '-' vs '_' but if we want to do something relatively immediate while we think of the deeper more long-term (but also much more complex, it sounds like) rules, then it looks like adding the script options as suffixes per fallback as a lookup option is a good first step?

nyurik commented:
i'm pretty sure there will be a lot of adjustments to be made, and very likely we will have to make a "language map" - geoshape with default language code, to simplify coordinate->lang code conversion.  Not very difficult code, but the actual generation of the map is tricky.

For now perhaps make it possible to specify exact codes in the config:   en. ja-Latn, ko_rm, etc, and if none matches, find a language that has the same script as the current one?

mooeypoo commented:
I'm really not in love with the idea of falling back on a somewhat random language that shares the same script, but I think it might be a decent first step for starters, especially since we'll be having that as a secondary fallback.

I'll revise the code.

nyurik commented on test/langPickerTest.js:
```diff
@@ -147,6 +147,45 @@ describe('LanguagePicker', () => {
+        { en: 'en value' },
+        { he: 'he value' },
+        { es: 'es value' },
+        { Cyrl: 'Cyrl value' },
```
here you should have  `ru: 'ru value'` instead of Cyrl. There is no such thing as "Cyrl" in OSM db.

mooeypoo committed:
Change script fallback to find language of same script

mooeypoo commented:
So I fixed the script, please recheck the tests and the way it works. I'm collecting the list of languages that use the same script, and am falling back on it.

However, and this is where my biggest concern is -- we have to limit the number of fallbacks because otherwise we'll try hundreds of languages, but the languages in the script fallback are not even remotely prioritized. This means that if we pick "en" we seem to not even remotely try "es" or "de" because they're waaaaaaay down the list, so we end up with, most likely, nothing.

However... it does seem to work on some languages, and seem to somewhat increase the odds of getting values.

I'm not sure how much I like this solution, but it might be better than nothing, unless we can find a way to prioritize the language-script list.

mooeypoo committed:
Fix tests for 10 language limit

nyurik commented:
I think we shouldn't limit to the first 10 - that seems arbitrary and won't result in useful results.  I think instead the fallbacks should be a string->int dictionary, so that you can quickly iterate through tags, getting an integer for each value, and using the one with the smallest value. What do you think?

mooeypoo committed:
Redesign the language fallback system

Fallback on languages now has the following rules, in order,
when checking and prioritizing existing values:

1. Requested language
2. Fallback language from fallbacks.json
3. Local language with suffix of same script as requested language
4. Any language with the same script as requested language
5. English
6. Any language with suffix Latn
7. Local language (name tag)
8. First value given

mooeypoo commented:
@nyurik After discussing the potential fallbacks and looking at existing data, I've completely revamped the way that fallbacks work. I think the rules are a lot better now, going by your suggestion but also adding in some common themes we've seen in the data. Take a look at the new tests I've added; the descriptions are a little convoluted but I wanted to make sure the process is clear. 

I think this makes a lot more sense, and has a much higher potential of showing us missing data.

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
+      // Get the best value from the best language fallback:
+      // 1. Requested language
+      if (this.values[this.userLang]) {
+        return this.values[this.userLang];
```
why double lookup?  create a var, and return if truthy.

nyurik replied:
missed this one

mooeypoo replied:
Whoops, thanks

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
+      // 2. Fallback language from fallbacks.json
+      for (let i = 0; i < this.fallbacks.length; i++) {
+        if (this.values[this.fallbacks[i]]) {
+          return this.values[this.fallbacks[i]];
```
same thing, and use [for-of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) loop for readability and to avoid hidden bugs.

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
+      // 4. Another language with the same script
+      for (let i = 0; i < this.languagesInScript.length; i++) {
+        if (this.values[this.languagesInScript[i]]) {
+          return this.values[this.languagesInScript[i]];
```
same thing

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
+      // Only do this if the script isn't already latin;
+      // otherwise, we've done it in step 3
+      for (let i = 0; i < valueLangCodes.length; i++) {
+        if (valueLangCodes[i].endsWith('Latn')) {
```
should be `'-Latn'`  (see https://taginfo.openstreetmap.org/search?q=Latn )

mooeypoo replied:
There are a couple of values there with _Latn... wouldn't it be safer to not have the dash/underscore at all?

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
-        firstChoiceValue : value;
+      // If nothing is found:
+      // - If there's a name tag, return it
+      // - IF there's no name tag, return first choice
```
capitalized IF

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
+          return this.values[valueLangCodes[i]];
+        }
       }
 
```
you need two more cases - ends with `'_rm'` https://taginfo.openstreetmap.org/search?q=_rm and `'_pinyin'`  https://taginfo.openstreetmap.org/search?q=pinyin  (they are good if name:en is missing)

mooeypoo replied:
Yes, we were discussing this -- I want to add it as a separate commit on top; my plan is this (see if you like it?) --

Add a script_aliases.json file, where we define something like this -

{
   "Latn": "ro", "rm"
}
etc

Then, have the code add a fallback that builds a proper lookup (with regexp to search the end with - and _ variances) for the aliases of language scripts. This way, we could relatively easily add aliases we've missed or need to adjust, and we won't have to rely on what *seems* like a standard in OSM and may change (I'm worried about those nonstandard standards...)

But since this whole thing is already a little convoluted (look at the tests! ;) I figured I will add this as a subsequent commit, and we can review that separately and make sure that works right.

What do you think?

nyurik replied:
ja_rm does not mean a language code, it means that this string is a romanized version of the Japanese string (transliteration).  See [osm docs](https://wiki.openstreetmap.org/wiki/Multilingual_names#Japan).  It seems that some regions just chose to use `-Latn`, and some tend to use `_rm`.  Historical, but that means we should treat them similarly.  Lookup table converting ro to Latn wouldn't help.  That's why I think you should simply go through each case one by one (just like you did), and since we now pretty much know them all (less than 10 of them total), we might as well spell them out.  No need to do automagic of "-" vs "_" -- they don't mix.

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
+      }
+
+      // 5. English
+      if (this.values[this.prefix ? `${this.prefix}en` : 'en']) {
```
same, and you also perform the same complex string concat twice.  Same fix though - `result = ...; if (result) return result;`

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
+      }
+    },
+    getResult: () => {
+      const valueLangCodes = Object.keys(this.values);
```
move this closer to usage - very likely it won't be needed.  (this is not pre-ES5 JS, where vars should be declared at the top. ES5+ is actually opposite - the closer `let` & `const` to the usage, the better)

mooeypoo replied:
Hmm good point, though declaring at the top is still recommended as far as I can see (also, eslint rule). I can declare it as a 'let' and assign below, though.

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
+      }
+
+      // 3. Any language with suffix of same script as requested language
+      for (let i = 0; i < valueLangCodes.length; i++) {
```
use [for of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) loop - more readable.

nyurik commented:
P.S. I started a discussion in [OSM mailing list](https://lists.openstreetmap.org/pipermail/tagging/2018-April/035855.html) about creating a map of languages - which territory has what prevailing language.  I think eventually this will be needed in cases like -- what's better:  `ja_rm` or `it` for an English speaker?  The answer is actually not trivial - if you are looking at Italy, it should be `it`, but if you are looking at Japan, than `ja_rm` is better.

mooeypoo committed:
Fix code comments and optimize for es6

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -67,7 +67,7 @@ LanguagePicker.prototype.newProcessor = function newProcessor() {
     },
     getResult: () => {
-      const valueLangCodes = Object.keys(this.values);
+      let val;
```
i meant don't declare it here either - because then the var has undefined value until assigned later.  The var declaration at the top is a leftover from the no `let/const` days (when it did make sense due to variable hoisting and related bugs) - see posts like [this](https://medium.com/@bluepnume/theres-no-need-to-define-all-javascript-vars-once-at-the-top-of-a-function-and-there-hasn-t-been-a66b31f21822).  So just move this line to where you assign it - it makes code more concise, and doesn't pollute the scope with vars until they are needed.

nyurik replied:
LOL, silly me, I commented too fast I think :)

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -76,43 +76,48 @@ LanguagePicker.prototype.newProcessor = function newProcessor() {
-          return this.values[valueLangCodes[i]];
+      const valueLangCodes = Object.keys(this.values);
+      for (const langCode of valueLangCodes) {
+        val = this.values[langCode];
```
why do you need to do a lookup and check it, if you know that every langCode must exist in `this.values` (because you just created it that way)?  I would rewrite:
```js
 for (const langCode of valueLangCodes) {
  if (langCode.endsWith(this.langScript)) {
    return this.values[langCode];
  }
}
```

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -76,43 +76,48 @@ LanguagePicker.prototype.newProcessor = function newProcessor() {
-      for (let i = 0; i < this.languagesInScript.length; i++) {
-        if (this.values[this.languagesInScript[i]]) {
-          return this.values[this.languagesInScript[i]];
+      for (const langCode of this.languagesInScript) {
```
same

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -76,43 +76,48 @@ LanguagePicker.prototype.newProcessor = function newProcessor() {
-        if (valueLangCodes[i].endsWith('Latn')) {
-          return this.values[valueLangCodes[i]];
+      for (const value of valueLangCodes) {
+        if (value && value.endsWith('Latn')) {
```
why would value ever be false?

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -76,43 +76,48 @@ LanguagePicker.prototype.newProcessor = function newProcessor() {
       // - If there's a name tag, return it
-      // - IF there's no name tag, return first choice
+      // - If there's no name tag, return first choice
       return this.nameTag && this.values[this.nameTag] ?
```
use `val` too, avoids double lookup.  There might be a JS engine optimization for it, but I don't know if its there.

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -76,43 +76,48 @@ LanguagePicker.prototype.newProcessor = function newProcessor() {
       // 5. English
-      if (this.values[this.prefix ? `${this.prefix}en` : 'en']) {
-        return this.values[this.prefix ? `${this.prefix}en` : 'en'];
+      val = this.values[this.prefix ? `${this.prefix}en` : 'en'];
```
i think it would also be good to make `this.prefix ? `${this.prefix}en` : 'en'` into a class-level constant -- no point in constructing a string for every tag - string concatenations are expensive when doing hundreds of millions of them.

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -1,16 +1,39 @@
+
+
+  // Store language script
+  this.langScript = scripts[lang] || 'Latn';
```
also, forgot to mention - prepend langScript with a `-` -- this would be more accurate.

nyurik replied:
P.S. I meant to add it after you collect languages below - so that when you do `endsWith()`, it works more accurately.  Just a nitpick really.

mooeypoo replied:
But even in the list you sent, there's a ja_Latn (https://taginfo.openstreetmap.org/search?q=Latn)

Should I still force it to look for -[script] ? Can we trust that structure?

nyurik replied:
There is no `ja_Latn`, check again :)  I already fixed those a few days ago, but they usually take ~24hrs to update taginfo. We don't want to handle one-off mistakes - people are very creative in how they mess up, we don't want to encourage it :)  Someone always fixes it, sooner or later.  So yes, always check `-[script]`, where script is "Latn", "Cyrl", etc, but keep in mind that we can treat `"_rm"` for a some (all?) languages as a sinonym to `"-Latn"`.

mooeypoo replied:
... I could have sworn I saw one... my eyes are deceiving me.

Okay, if we can (for the most part) trust the structure, I'll add that to the "endsWith". Thanks!

nyurik replied:
I believe you  - it was there - but I fixed it two days ago I think. :)

mooeypoo replied:
I thought I was hallucinating... :smiley:

stephanebisson commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,88 @@ function LanguagePicker(lang = 'en', config = {}) {
-        (bestChoiceLang && fallbacks[bestChoiceLang]) ||
-        firstChoiceValue,
+    // This is for testing purposes
+    getFallbacks: () => this.fallbacks,
```
I'm looking at the tests but I'm not sure where this is being used...

mooeypoo replied:
It was on a previous iteration, I forgot to remove this

mooeypoo committed:
Fix per review comments

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -101,29 +101,26 @@ LanguagePicker.prototype.newProcessor = function newProcessor() {
-        this.values[this.nameTag] :
-        firstChoiceValue;
+      val = this.values[this.nameTag];
+      return this.nameTag && val ? val : firstChoiceValue;
```
you are using `this.nameTag` before you check if it actually has a value.  How about
```js
if (this.nameTag) {
  val = this.values[this.nameTag];
}
return val || firstChoiceValue;
```

mooeypoo replied:
ha! Missed that. 

Or, I could do 

```
return ( this.nameTag && this.values[this.namesTag] ) || firstChoiceValue;
```

Which is the same as yours, just short-formed.

mooeypoo committed:
Add dash prefix to scripts; fix code style

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -87,7 +87,7 @@ LanguagePicker.prototype.newProcessor = function newProcessor() {
       const valueLangCodes = Object.keys(this.values);
       for (const langCode of valueLangCodes) {
-        if (langCode.endsWith(this.langScript)) {
+        if (langCode.endsWith(`-${this.langScript}`)) {
```
almost there, but a few minor things:   this is not very efficient - you concat a string that doesn't change (same as the prefixed en before)

mooeypoo replied:
Good point; Doing that for the next commit.

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,84 @@ function LanguagePicker(lang = 'en', config = {}) {
+      // 3. Any language with suffix of same script as requested language
+      const valueLangCodes = Object.keys(this.values);
+      for (const langCode of valueLangCodes) {
+        if (langCode.endsWith(`-${this.langScript}`)) {
```
this should be a class-level var to avoid repeated string concatenation on each pass

mooeypoo committed:
Caching lang code template

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,86 @@ function LanguagePicker(lang = 'en', config = {}) {
+      // 6. Any language with suffix 'Latn'
+      // Only do this if the script isn't already latin;
+      // otherwise, we've done it in step 3
+      for (const langCode of valueLangCodes) {
```
thanks to your comment, this for loop should be inside an if statement - to skip it if `prefixedLangScript === '-Latn'`

mooeypoo replied:
Wasn't.... it? ... I think I took it out in one of the edits. Whoops

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,86 @@ function LanguagePicker(lang = 'en', config = {}) {
-        firstChoiceValue : value;
+      // If nothing is found:
+      // - If there's a name tag, return it
+      // - If there's no name tag, return first choice
```
sorry, a tiny nitpick - could you change the comment to " Otherwise, return first choice". I keep misreading it because it only differs in two letters with previous line.

mooeypoo replied:
Yup, fair. Changing.

nyurik commented on lib/LanguagePicker.js:
```diff
@@ -19,50 +42,86 @@ function LanguagePicker(lang = 'en', config = {}) {
+          return this.values[langCode];
+        }
       }
 
```
So just to double check - you don't want to include _rm suffix search here for now, in the same way as you did -Latn search?  Because there are four times as many "_rm" suffixes (~260,000) as there are "-Latn" (~75,000), so if anything, we should probably process _rm only.  But _rm tends to be clustered much more in Asia, especially in Korea, whereas -Latn is more global - that's why i think we should do both, and _rm should go after -Latn.

mooeypoo replied:
hmm... is _rm the same idea as -Latn ? My worry was anything other than _rm, like, generalizing the script, but you're right that in this particular fallback we're using "Latn" anyways, so we might as well?

I would like to have something that also generalizes the process for other potential codes. Is there anything other than _rm and _ro that we should consider for other scripts (possibly in a followup) ?

mooeypoo committed:
Add _rm and zh_pinyin checks to -Latn fallback

mooeypoo committed:
Split -Latn and _rm/zh_pinyin searches to two iterations

nyurik commented:
You rock :)
